lib/semantics/tools.h

```diff
@@ -160,5 +160,18 @@ template<typename T> const SomeExpr *GetExpr(const T &x) {
   return GetExprHelper{}.Get(x);
 }
 
+template<typename T> std::optional<std::int64_t> GetIntValue(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    return evaluate::ToInt64(*expr);
+  } else {
+    return std::nullopt;
+  }
+}
+
+template<typename T> std::optional<std::int64_t> GetPosIntValue(const T &x) {
+  const auto v{GetIntValue(x)};
```

Comment:
This would be more readable as

```
if (const auto v{GetIntValue(x)}) {
  if (*v > 0) {
    return v;
  }
}
return std::nullopt;
```

(or at least I think it would be)

I just noticed that you don't consider 0 to be positive.  That might confuse some expectations of readers who think that "positive" means "not negative".  So `GetPosIntValue` is maybe not a good name.  Try something like `GetIntValueIfOverZero`.

Reply:
I'm not sure this function is useful anyway. It might be easier just to use `GetIntValue()` and compare the result to `0` when you need to.

Reply:
Thank you both for the suggestions. I have got rid of this `GetPosIntValue` function in the latest commit. All the checks will analyze the result from `GetIntValue` now.